[OSDOCS#4530] etcd performance mod updates for CL (#60453)
openshift/openshift-docs: modules/recommended-etcd-practices.adoc

Those latencies can cause etcd to miss heartbeats, not commit new proposals to the disk on time, and ultimately experience request timeouts and temporary leader loss. High write latencies also lead to an OpenShift API slowness, which affects cluster performance. Because of these reasons, avoid colocating other workloads on the control-plane nodes that are I/O sensitive or intensive and share the same underlying I/O infrastructure. 

In terms of latency, run etcd on top of a block device that can write at least 50 IOPS of 8000 bytes long sequentially. That is, with a latency of 10ms, keep in mind that uses fdatasync to synchronize each write in the WAL. For heavy loaded clusters, sequential 500 IOPS of 8000 bytes (2 ms) are recommended. To measure those numbers, you can use a benchmarking tool, such as fio.

To achieve such performance, run etcd on machines that are backed by SSD or NVMe disks with low latency and high throughput. Consider single-level cell (SLC) solid-state drives (SSDs), which provide 1 bit per memory cell, are durable and reliable, and are ideal for write-intensive workloads.

> NOTE: The load on etcd arises from static factors, such as the number of nodes and pods, as well as dynamic factors, including changes in endpoints due to horizontal pod autoscaling, pod restarts, job executions, and other workload-related events. To accurately size your etcd setup, you must analyze the specific requirements of your workload. Consider the number of nodes, pods, and other relevant factors that impact the load on etcd.

Reviewer: This note does not follow or standard note syntax. PLease convert the note to follow the following syntax:

> [NOTE]
====
[Text]
====

See, [Admonitions]https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#admonitions)

Author: It's [standard adoc](https://docs.asciidoctor.org/asciidoc/latest/blocks/admonitions/#admonition-syntax), but no skin off my nose. 

```suggestion
[NOTE]
====
The load on etcd arises from static factors, such as the number of nodes and pods, as well as dynamic factors, including changes in endpoints due to horizontal pod autoscaling, pod restarts, job executions, and other workload-related events. To accurately size your etcd setup, you must analyze the specific requirements of your workload. Consider the number of nodes, pods, and other relevant factors that impact the load on etcd.
====
```

> NOTE: The load on etcd arises from static factors, such as the number of nodes and pods, as well as dynamic factors, including changes in endpoints due to horizontal pod autoscaling, pod restarts, job executions, and other workload-related events. To accurately size your etcd setup, you must analyze the specific requirements of your workload. Consider the number of nodes, pods, and other relevant factors that impact the load on etcd.

Reviewer: Lot so noun listed in one sentence. Because a note is a self-contained entity, could the nouns be placed in a list?

As well as > and

Those latencies can cause etcd to miss heartbeats, not commit new proposals to the disk on time, and ultimately experience request timeouts and temporary leader loss. High write latencies also lead to an OpenShift API slowness, which affects cluster performance. Because of these reasons, avoid colocating other workloads on the control-plane nodes that are I/O sensitive or intensive and share the same underlying I/O infrastructure. 

In terms of latency, run etcd on top of a block device that can write at least 50 IOPS of 8000 bytes long sequentially. That is, with a latency of 10ms, keep in mind that uses fdatasync to synchronize each write in the WAL. For heavy loaded clusters, sequential 500 IOPS of 8000 bytes (2 ms) are recommended. To measure those numbers, you can use a benchmarking tool, such as fio.

To achieve such performance, run etcd on machines that are backed by SSD or NVMe disks with low latency and high throughput. Consider single-level cell (SLC) solid-state drives (SSDs), which provide 1 bit per memory cell, are durable and reliable, and are ideal for write-intensive workloads.

[NOTE]
====

> The load on etcd arises from static factors, such as the number of nodes and pods, and dynamic factors, including changes in endpoints due to horizontal pod autoscaling, pod restarts, job executions, and other workload-related events. To accurately size your etcd setup, you must analyze the specific requirements of your workload. Consider the number of nodes, pods, and other relevant factors that impact the load on etcd.

Reviewer 2: Can this be ~~horizontal~~ pod autoscaling, as there are multiple ways to autoscale pods?

Author: ```suggestion
The load on etcd arises from static factors, such as the number of nodes and pods, and dynamic factors, including changes in endpoints due to pod autoscaling, pod restarts, job executions, and other workload-related events. To accurately size your etcd setup, you must analyze the specific requirements of your workload. Consider the number of nodes, pods, and other relevant factors that impact the load on etcd.
```

* RAID 0 technology for increased performance.
* Dedicated etcd drives. Do not place log files or other heavy workloads on etcd drives.

> Avoid NAS or SAN setups, spinning drives, and software-defined storage, like Ceph. Always benchmark by using utilities such as fio. Continuously monitor the cluster performance as it increases.

Author: @mandre @pierreprinetti How about something like this?

```suggestion
Avoid NAS or SAN setups and spinning drives. Ceph RBD, as well as other network-attached storage, can result in unpredictable network latency. If possible, use PCI passthrough with NVM devices instead. 

Always benchmark by using utilities such as fio. Continuously monitor the cluster performance as it increases. 
```

> Avoid NAS or SAN setups, spinning drives, and software-defined storage, like Ceph. Always benchmark by using utilities such as fio. Continuously monitor the cluster performance as it increases.

Reviewer 3: Yes!

* RAID 0 technology for increased performance.
* Dedicated etcd drives. Do not place log files or other heavy workloads on etcd drives.

> Avoid NAS or SAN setups and spinning drives. Ceph RBD, as well as other network-attached storage, can result in unpredictable network latency. If possible, use PCI passthrough with NVM devices instead. 

Reviewer: Spell out NAS, SAN, RBD, PCI, and NVM, on first occurrence. 

I do not think are well-known acronyms or abbreviations.

Reviewer: As well as > and

![Screenshot from 2023-06-09 14-38-46](https://github.com/openshift/openshift-docs/assets/57954076/ed7f6b4e-8683-42ff-8681-23ab14b81061)

Reviewer: > If possible, use PCI passthrough with NVM devices instead.

"Instead" seems to dangle or be misplaced with the priority it's trying to convey. How about:

> If possible, use PCI passthrough with NVM devices instead of other X.

I'm not sure what generic components are being compared.

Reviewer: I also think that this paragraph reads as a note and should be enclosed in admonition syntax. WDYT?

Author: I'd beg to differ about most of those acronyms given that our audience consists of people who are looking to optimize etcd performance. In my judgment, though, happy to spell out RBD. 

```suggestion
[NOTE]
====
Avoid NAS or SAN setups and spinning drives. Ceph Rados Block Device (RBD), as well as other network-attached storage, can result in unpredictable network latency. To provide fast storage to etcd nodes at scale, use PCI passthrough to pass NVM devices directly to the nodes. 
====
```

> Always benchmark by using utilities such as fio. Continuously monitor the cluster performance as it increases. 

Reviewer: Surround tool in backticks? (I see an example in the OCP guidelines doc)

> Always benchmark by using utilities such as `fio`. 

> Continuously monitor the cluster performance as it increases.

What is doing this monitoring, the tool or a user? I think the sentence needs a subject is missing. 

![Screenshot from 2023-06-09 14-43-43](https://github.com/openshift/openshift-docs/assets/57954076/2c460a63-a45f-4882-a4f6-a67d9d5c7ad9)

Author: "fio" is used as-is in the live docs already. I'm apt to leave it simply because I don't work in the scale and perf modules much. 

```suggestion
Always benchmark by using utilities such as fio. You can use such utilities to continuously monitor the cluster performance as it increases. 
```

I'm okay with the imperative statement, but also happy to be explicit.

* RAID 0 technology for increased performance.
* Dedicated etcd drives. Do not place log files or other heavy workloads on etcd drives.

[NOTE]
====

> Avoid NAS or SAN setups and spinning drives. Ceph Rados Block Device (RBD), as well as other network-attached storage, can result in unpredictable network latency. To provide fast storage to etcd nodes at scale, use PCI passthrough to pass NVM devices directly to the nodes. 

Reviewer: One more nit from my side:

as well as > and

![Screenshot from 2023-06-27 10-53-53](https://github.com/openshift/openshift-docs/assets/57954076/90a757d5-07e9-41e8-af47-d2d05aa35aee)

Author: Feel free to just use the GH suggestion feature here, but yeah, changing.

```suggestion
Avoid NAS or SAN setups and spinning drives. Ceph Rados Block Device (RBD) and other types of network-attached storage can result in unpredictable network latency. To provide fast storage to etcd nodes at scale, use PCI passthrough to pass NVM devices directly to the nodes. 
```
